Added changes to allow db:status to determine if setup:upgrade is req…

BarnyShergold opened the pull request:
### Description
During the Deploy Phase - Upgrade/Setup ALWAYS runs setup:upgrade.
It has been determined that the time overhead of running setup:db:status was average 5s but setup:upgrade was average 14s. This check will reduce the deploy phase by 11s

These tests were done on a Sandbox Pro-12 Environment - project ID z3ek236yy2yv2 and on the STAGING branch.

Whilst this will add up to 5s to a deploy that does require DB-Updates - these SHOULD be a) run separate to functional deploys b) be a much smaller percentage of deploys.

On the environment above the downtime for deploy was clocked at 13s (Maintenance On to Maintenance Off) this means sessions requests only being queued for that time. This is much more desirable than the previous 24s

### Fixed Issues (if relevant)
1. magento/ece-tools#419: [ECE-Tools] Conditionally Execute setup:upgrade During Deployment

### Manual testing scenarios
1. Generate a build that runs bin/magento setup:upgrade
2. SSH into server and run bin/magento setup:db:status
3. If DB reports that upgrade STILL needs running then ensure patch MDVA-21578_EE_2.3.2_COMPOSER_v1.patch is installed on project and repeat 1-3
4. Generate a build that requires a DB update - check logs - setup:db:status should be logged and then the setup:upgraded immediately after.
5. Generate a build that DOES NOT require a DB update - check logs - setup:db:status should be logged and then the setup:upgrade will be absent.
6. REPEAT STEP 4

### Contribution checklist
 - [X ] Pull request has a meaningful description of its purpose
 - [X ] All commits are accompanied by meaningful commit messages
 - [NA ] All new or changed code is covered with unit/integration tests (if applicable)
 - [NA ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Added changes to allow db:status to determine if setup:upgrade is required

BarnyShergold commented:
@YPyltiai - This is my first ever contribution so I am not sure what Travis CI is and why it's failed

kandy commented on src/Step/Deploy/InstallUpdate/Update/Setup.php:
```diff
@@ -92,22 +94,41 @@ public function __construct(
+            }
+        }
+
+        if ($outdated) {
```
We change default behavior of upgrade that can cause issues.I suggest making this functionality configurable and disabled by default

BarnyShergold replied:
So perhaps adding something like :
```
# SKIP_SETUP_UPGRADE - skips setup:upgrade during the deploy phase if not required.                                           #
# Magento Version: 2.3.3 and later                                                                                    #
# Default value: false                                                                                              #
# Possible values: true or false                                                                                      #
# Stages: deploy                                                                                    #
# Example:                                                                                                            #
#          stage:                                                                                                     #
#            deploy:                                                                                                   #
#              SKIP_SETUP_UPGRADE: true       
```
In .magento.env.yaml ?

kandy replied:
Yep
Also I propose to name it more understandable, like
skip_upgrade_unless_required

BarnyShergold replied:
Sure!
I'm sure I can work out how to do this. I guess because of this I will need to add at least one Unit Test to Magento\MagentoCloud\Test\Unit\Step\Deploy\InstallUpdate\Update\SetupTest .

I have a question - as the code structure of ECE-TOOLS is different in the repo - are there some guidelines somewhere how run locally to test? Both the actual code and the UTs?

YPyltiai replied:
@BarnyShergold you can run all functional test using Cloud Docker: https://devdocs.magento.com/guides/v2.3/cloud/docker/docker-development-testing.html

Unit tests can be run directly I think.

BarnyShergold replied:
Hi - Okay that might be fun then - I don't use Docker - Vagrant instead.

YPyltiai replied:
@BarnyShergold , this actually just requires Docker itself to be installed - then following documentation everything will be deployed for you automatically.

BarnyShergold replied:
I've tried that but getting errors all over the place. What's the best place to go to discuss this and get some help?

YPyltiai replied:
@BarnyShergold , please join https://magentocommeng.slack.com/ channel - #cloud. We can discuss it over Slack there.

Thanks!

arhiopterecs replied:
@BarnyShergold 
Thanks so much for your pull request.
Are you working on this PR?
If you want, we can finish this PR on our side.

BarnyShergold replied:
Hi Had to put to one side for now but was going to follow Kandy's suggestion and make it configurable - will try to pick up in next few days

BarnyShergold replied:
@arhiopterecs - All code changes made & commit with extensive Unit Tests added. Not sure why Travis is failing but it looks external to this module

shiftedreality commented:
Internal task: https://magento2.atlassian.net/browse/MAGECLOUD-4389

andriyShevtsov commented:
One more previously created internal task https://magento2.atlassian.net/browse/MAGECLOUD-3183

Commit:
Added code to allow configuration option SKIP_UPGRADE_UNLESS_REQUIRED
Added definition to .magenot.env.yaml
Added PHPUnit Tests (includes updating existing ones)

Commit:
Merge branch 'develop' into develop

Commit:
Merge branch 'develop' into develop

shiftedreality commented:
Putting `on hold` whole discussions are going

shiftedreality commented:
Hello @BarnyShergold,

@arhiopterecs measured the deployment time and wanted to share those results with you:

```
Execution time: ~ 22 sec
Clean cache: ~ 4,3 sec
Generate new config cache: ~ 5-7 sec
Bypass all modules and check for updates: ~ 0,34 sec
Run all recurring scripts: ~ 12 sec;
Total: ~ 21- 23 sec
```

Even running recurring scripts only, would still require regeneration of caches so the value of this option will be eliminated. We propose to close this issue and possibly proceed with some internal Magento improvements if any.

Thank you for this suggestion, if you will have any more ideas - please feel free to re-open this issue.

cc: @YPyltiai, @kandy

ghost commented:
Hi @BarnyShergold, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento+ece-tools+619), it will take less than a minute. </br>Your feedback will help us to improve contribution process.
